agent/runner/runner.go

```diff
@@ -48,30 +53,100 @@ type Runner struct {
 	actionsMessages chan agentpb.AgentRequestPayload
 	jobsMessages    chan agentpb.AgentResponsePayload
 
-	sem *semaphore.Weighted
-	wg  sync.WaitGroup
+	wg sync.WaitGroup
 
-	rw      sync.RWMutex
-	rCancel map[string]context.CancelFunc
+	// cancels holds cancel functions for running actions and jobs.
+	cancelsM sync.RWMutex
+	cancels  map[string]context.CancelFunc
+
+	// running holds IDs of running actions and jobs.
+	runningM sync.RWMutex
+	running  map[string]struct{}
+
+	// gSem is a global semaphore to limit total number of concurrent operations performed by the runner.
+	gSem *semaphore.Weighted
+
+	// tokenCapacity is a limit of concurrent operations on a single resource, usually database instance.
+	tokenCapacity uint16
+
+	// lSems is a map of local semaphores to limit number of concurrent operations on a single database instance.
+	// Key is a token which is typically is a hash of DSN(only host:port pair), value is a semaphore.
+	lSemsM sync.Mutex
+	lSems  map[string]*entry
+}
+
+// entry stores local semaphore and its counter.
+type entry struct {
+	count atomic.Int32
+	sem   *semaphore.Weighted
 }
 
 // New creates new runner. If capacity is 0 then default value is used.
-func New(capacity uint16) *Runner {
+func New(totalCapacity, tokenCapacity uint16) *Runner {
 	l := logrus.WithField("component", "runner")
-	if capacity == 0 {
-		capacity = defaultCapacity
+	if totalCapacity == 0 {
+		totalCapacity = defaultTotalCapacity
+	}
+
+	if tokenCapacity == 0 {
+		tokenCapacity = defaultTokenCapacity
 	}
 
-	l.Infof("Runner capacity set to %d.", capacity)
+	l.Infof("Runner capacity set to %d, token capacity set to %d", totalCapacity, tokenCapacity)
 
 	return &Runner{
 		l:               l,
 		actions:         make(chan actions.Action, bufferSize),
 		jobs:            make(chan jobs.Job, bufferSize),
-		sem:             semaphore.NewWeighted(int64(capacity)),
-		rCancel:         make(map[string]context.CancelFunc),
+		cancels:         make(map[string]context.CancelFunc),
+		running:         make(map[string]struct{}),
 		jobsMessages:    make(chan agentpb.AgentResponsePayload),
 		actionsMessages: make(chan agentpb.AgentRequestPayload),
+		tokenCapacity:   tokenCapacity,
+		gSem:            semaphore.NewWeighted(int64(totalCapacity)),
+		lSems:           make(map[string]*entry),
+	}
+}
+
+// acquire acquires global and local semaphores.
+func (r *Runner) acquire(ctx context.Context, token string) error {
+	if token != "" {
+		r.lSemsM.Lock()
+
+		e, ok := r.lSems[token]
+		if !ok {
+			e = &entry{sem: semaphore.NewWeighted(int64(r.tokenCapacity))}
+			r.lSems[token] = e
+		}
+		r.lSemsM.Unlock()
+
+		if err := e.sem.Acquire(ctx, 1); err != nil {
+			return err
+		}
+		e.count.Add(1)
+	}
+
+	if err := r.gSem.Acquire(ctx, 1); err != nil {
+		return err
```

Comment:
shouldn't we release from the semaphore above in case of error here?

Reply:
Yes, good catch. Fixed.